Improve SpriteFrameEditor frame addition ordering

AThousandShips opened the pull request:
See the original proposal, implements drawing the index of selected frames and the modes suggested

![image](https://user-images.githubusercontent.com/96648715/199310344-4c51901a-df5b-48e4-84fc-ee6a45c5636b.png)

![image](https://user-images.githubusercontent.com/96648715/199310244-b11f3df5-6c7c-48d1-bb5e-375e0f42668e.png)

Fixes: [godot-proposals#5679](https://github.com/godotengine/godot-proposals/issues/5679)

kleonc commented on editor/plugins/sprite_frames_editor_plugin.cpp:
```diff
@@ -299,21 +320,103 @@ void SpriteFramesEditor::_sheet_zoom_reset() {
+
+			const int neg_frame_x = frame_count.x - (pos_frame_x + 1);
+			const int neg_frame_y = frame_count.y - (pos_frame_y + 1);
+
+			switch (frame_order) {
+				case FRAME_ORDER_LEFT_RIGHT_TOP_BOTTOM: {
+					to_pair.first = frame_count.x * pos_frame_y + pos_frame_x;
+				} break;
+
+				case FRAME_ORDER_LEFT_RIGHT_BOTTOM_TOP: {
+					to_pair.first = frame_count.x * neg_frame_y + pos_frame_x;
+				} break;
+
+				case FRAME_ORDER_RIGHT_LEFT_TOP_BOTTOM: {
+					to_pair.first = frame_count.x * pos_frame_y + neg_frame_x;
+				} break;
+
+				case FRAME_ORDER_RIGHT_LEFT_BOTTOM_TOP: {
+					to_pair.first = frame_count.x * neg_frame_y + neg_frame_x;
+				} break;
+
+				case FRAME_ORDER_TOP_BOTTOM_LEFT_RIGHT: {
+					to_pair.first = pos_frame_y + frame_count.y * pos_frame_x;
+				} break;
+
+				case FRAME_ORDER_TOP_BOTTOM_RIGHT_LEFT: {
+					to_pair.first = pos_frame_y + frame_count.y * neg_frame_x;
+				} break;
+
+				case FRAME_ORDER_BOTTOM_TOP_LEFT_RIGHT: {
+					to_pair.first = neg_frame_y + frame_count.y * pos_frame_x;
+				} break;
+
+				case FRAME_ORDER_BOTTOM_TOP_RIGHT_LEFT: {
+					to_pair.first = neg_frame_y + frame_count.y * neg_frame_x;
+				} break;
+			}
+
+			// Assign in vector.
+			frames_ordered.set(index++, to_pair);
```
Might be clearer:
```suggestion
			const int idx = from_pair.key;
			const int selection_order = from_pair.value;

			int order_by = selection_order;

			// Extract coordinates for sorting.
			const int pos_frame_x = idx % frame_count.x;
			const int pos_frame_y = idx / frame_count.x;

			const int neg_frame_x = frame_count.x - (pos_frame_x + 1);
			const int neg_frame_y = frame_count.y - (pos_frame_y + 1);

			switch (frame_order) {
				case FRAME_ORDER_LEFT_RIGHT_TOP_BOTTOM: {
					order_by = frame_count.x * pos_frame_y + pos_frame_x;
				} break;

				case FRAME_ORDER_LEFT_RIGHT_BOTTOM_TOP: {
					order_by = frame_count.x * neg_frame_y + pos_frame_x;
				} break;

				case FRAME_ORDER_RIGHT_LEFT_TOP_BOTTOM: {
					order_by = frame_count.x * pos_frame_y + neg_frame_x;
				} break;

				case FRAME_ORDER_RIGHT_LEFT_BOTTOM_TOP: {
					order_by = frame_count.x * neg_frame_y + neg_frame_x;
				} break;

				case FRAME_ORDER_TOP_BOTTOM_LEFT_RIGHT: {
					order_by = pos_frame_y + frame_count.y * pos_frame_x;
				} break;

				case FRAME_ORDER_TOP_BOTTOM_RIGHT_LEFT: {
					order_by = pos_frame_y + frame_count.y * neg_frame_x;
				} break;

				case FRAME_ORDER_BOTTOM_TOP_LEFT_RIGHT: {
					order_by = neg_frame_y + frame_count.y * pos_frame_x;
				} break;

				case FRAME_ORDER_BOTTOM_TOP_RIGHT_LEFT: {
					order_by = neg_frame_y + frame_count.y * neg_frame_x;
				} break;
			}

			// Assign in vector.
			frames_ordered.set(index++, Pair<int, int>(order_by, idx));
```

kleonc commented on editor/plugins/sprite_frames_editor_plugin.cpp:
```diff
@@ -299,21 +320,103 @@ void SpriteFramesEditor::_sheet_zoom_reset() {
+			frames_ordered.set(index++, to_pair);
+		}
+
+		// Sort frames
```
```suggestion
		// Sort frames.
```

kleonc reviewed (commented):
Not sure if it's just me but to me the horizontal/vertical part of the orderings seem inverted, e.g. I'd expect _"Left to Right, Top to Bottom"_ to go from left to right first in a column by column manner, and then top to bottom for each such column. But it currently goes row by row from top to bottom and then left to right for each such row (which I think _"Top to Bottom, Left to Right"_ should be doing). :thinking:

It might be not clear what the added dropdown is for. Especially when no frames are selected and thus changing selected ordering doesn't affect anything. Thus probably a `Label` before the dropdown is needed, just like for the other options.

Another thing is this dialog is getting way too wide because of that many options stacked in a single horizontal line. I suggest moving these into a `HFlowContainer`, grouped appropriately into `HBoxContainer`s (so contents of each such `HBoxContainer`s would stay together):
![AL8GbVSVxO](https://user-images.githubusercontent.com/9283098/199098251-4f8ca939-646b-403e-a7a3-7eb8b9e40628.png)

Overall LGTM. Code looks fine, seems to be working as intended. The usage of map/pair might make the code a little confusing but there's not much of it so it should be fine (few more comments for clarification should do the thing).

Also maybe the numbers/text should not be drawn when zoomed out enough. The text can be bigger than the frame at a certain zoom.

AThousandShips commented:
Will take a look at some of the visual details and clean up some code

Would "Left to right, then top to bottom" be more clear? I personally read "Left to right, top to bottom" as "left to right in each row, then top to bottom", can implement the original suggestion for separate modes for horizontal and vertical and will look into this

kleonc commented:
> Would "Left to right, then top to bottom" be more clear?

I don't think it would change anything, "then" is kinda implicitly in there already. I mean with/without it I think the interpretation would be the same (whatever the interpretation is).

>I personally read "Left to right, top to bottom" as "left to right in each row, then top to bottom", , i.e. row major order,

Yeah, I can see how it can be interpreted in a different way. That's why I've started my comment with _"Not sure if it's just me (...)"_. Maybe I'm in the minority in here, I don't know. :slightly_smiling_face:

> can implement the original suggestion for separate modes for horizontal and vertical and will look into this

If you're referring to my original suggestion (up to 3 dropdowns) then I don't think it would be good. A single dropdown is way easier to use. I considered also submenus (another dropdown from the dropdown) but it would be harder to use and it's not like there are many options to choose from so there's no need to categorize like that. Maybe just adding some text into preceding separators could help? :thinking: (you can pass a text to `OptionButton::add_separator`)

AThousandShips commented:
Oh didn't know about separators as text, that I will look into! Thank you

AThousandShips commented:
Fixed the various style and code suggestions, added clarifications to the drop down hopefully that makes it clearer, fixed the width of the popup, and made the indices stop rendering when too large

@kleonc thank you for your feedback, anything else you can think of?

kleonc commented:
> anything else you can think of?

Frame numbers text needs a different color border or something like that, currently it can be not visible depending on the background color. You could probably look up how it's solved e.g. in the `TileMap` editor.

Regarding the `HFlowContainer`:
- `VSeparator`s between `HBoxConatiner`s don't look good in this setup, especially when they move between the rows:
![CnzgNDNzFp](https://user-images.githubusercontent.com/9283098/199307589-e23b6124-0fd3-4210-9eca-cd219ce7f33b.gif)
- The rows are not separated visually anyhow, making the groups (`HBoxConatiner`s) not easily distinguishable.

I think what could help is to remove all these `VSeparator`s and move each group (`HBoxContainer`) to a `PanelContainer` (or something like that) so each such group would have its own darker background separating it from the surroundings (both horizontally and vertically).

Also maybe some fixed minimum size should be set on this dialog to disallow the options to be clamped like in the GIF above. Personally I do like I can shrink it as much as possible but maybe it's not "visually pleasant" or something like that. Not sure, I'm not a "GUI guy". :smile:

kleonc reviewed (approved):
Approving functionality / code, LGTM.
And works fine as far as I've tested.

Leaving the visuals / presentation / wording etc. for the others. :slightly_smiling_face:

AThousandShips commented:
@kleonc would you mind pointing some relevant code owners to this, I don't know how to and who would be appropriate

kleonc commented:
> @kleonc would you mind pointing some relevant code owners to this, I don't know how to and who would be appropriate

cc @godotengine/2d-editor I guess.

You could for example ask for a review in [the Contributors Chat](https://chat.godotengine.org) (for this one in [the editor channel](https://chat.godotengine.org/channel/editor)?). And in general you can ask/discuss things in there.

AThousandShips commented:
Thank you will check that!

AThousandShips commented:
Testing an alternative layout:
![image](https://user-images.githubusercontent.com/96648715/199555209-1d850eec-4fe4-4382-92bb-27a6523b6004.png)

hsandt commented:
Nice, I didn't test it yet but the screenshots seem to match what I was proposing.

I also get the ambiguity about Left-to-right, Top-to-bottom and reversely. This seems to be the official wording for it, but for users not used to it, I think the categories "By row" and "By column" should make the distinction. The user can play around with the settings the first time and observe the numbers to understand the differences. That said, to make the items clearer there are a few possibilities:

- we could be more explicit e.g. displaying "Left to right inside a row, Top to bottom rows" or "Left to right inside a row, rows navigated Top to bottom". But it will get longer so that's a compromise to do
- we could add an icon showing the navigation order (e.g. a Z-shaped arrow for Left to right, Top to bottom). I don't know how easy or hard it is to add an icon like this inside or besides text, and I wouldn't expect it for this PR, but it could be a future improvement.

AThousandShips commented:
@hsandt thank you for you feedback, I'll look into icons!

KoBeWi commented on editor/plugins/sprite_frames_editor_plugin.h:
```diff
@@ -107,9 +124,14 @@ class SpriteFramesEditor : public HSplitContainer {
+	HashMap<int, int> frames_selected; // Key is frame index. Value is selection order.
 	HashSet<int> frames_toggled_by_mouse_hover;
+	Vector<Pair<int, int>> frames_ordered; // First is the index to be ordered by. Second is the actual frame index.
+	int selected_count = 0;
```
What is `selected_count` for? Isn't it the same as `frames_selected.size()`?

AThousandShips replied:
Its for sorting by selected order, if frames are removed and new added the size of the map does not apply

KoBeWi replied:
I tested the `frames_selected.size()` and it seems to work the same. What exactly could break?

AThousandShips replied:
It'd take about 68 years if you select one on average every second, just for reference

However I think a different solution will actually work will try it out tomorrow

hsandt commented:
OK, I tested it, I think it's pretty clear even without icons since there are headers "By Row" and "By Column" so we cannot be confused on which one is iterated on first.

The selection is also refreshed when changing ordering mode, for immediate feedback, so it's very convenient to use.

![image](https://user-images.githubusercontent.com/1067772/219867237-52d801ab-6812-466c-8f50-2082f3a02e64.png)

The only question is whether users will think about revealing the pane on the right when hidden, but it's likely that they do when the slicing is not correct anyway. Plus, I think it was revealed by default, so users have a chance to see it once before they hide it the first time.

![image](https://user-images.githubusercontent.com/1067772/219867341-64e9d3c7-1db7-41ec-8957-274d9d24d788.png)

I also see it was approved by reviewer in the meantime, so it should be good to go!

AThousandShips committed:
Improve SpriteFrameEditor frame addition ordering

KoBeWi commented on editor/plugins/sprite_frames_editor_plugin.h:
```diff
@@ -38,6 +38,7 @@
 #include "scene/gui/dialogs.h"
 #include "scene/gui/item_list.h"
 #include "scene/gui/line_edit.h"
+#include "scene/gui/option_button.h"
```
This can be forward-declared and moved to cpp.

AThousandShips replied:
Will take a look tomorrow! Thank you

AThousandShips commented:
I'll go ahead and squash in a little bit if there's no questions about the interface

YuriSizov commented on editor/plugins/sprite_frames_editor_plugin.cpp:
```diff
@@ -209,13 +228,18 @@ void SpriteFramesEditor::_sheet_preview_input(const Ref<InputEvent> &p_event) {
 				frames_selected.erase(idx);
 			} else {
-				frames_selected.insert(idx);
+				frames_selected.insert(idx, selected_count++);
```
I'd split this line so that the logic of using and incrementing the property is clear.

YuriSizov commented on editor/plugins/sprite_frames_editor_plugin.cpp:
```diff
@@ -300,21 +327,106 @@ void SpriteFramesEditor::_sheet_zoom_reset() {
 		if (!frames_selected.has(i)) {
-			frames_selected.insert(i);
-			should_clear = false;
+			frames_selected.insert(i, selected_count++);
```
Same here.

YuriSizov commented on editor/plugins/sprite_frames_editor_plugin.cpp:
```diff
@@ -300,21 +327,106 @@ void SpriteFramesEditor::_sheet_zoom_reset() {
+		return;
+	}
+
+	frames_need_sort = false;
+
+	frames_ordered.resize(frames_selected.size());
+
+	if (frames_selected.is_empty()) {
+		return;
+	}
+
+	const Size2i frame_count = _get_frame_count();
+
+	const int frame_order = split_sheet_order->get_selected_id();
+
+	int index = 0;
+
+	// Fill based on order.
```
Nit: There are too many empty lines here, it will probably look cleaner with some light grouping.

YuriSizov commented on editor/plugins/sprite_frames_editor_plugin.cpp:
```diff
@@ -300,21 +327,106 @@ void SpriteFramesEditor::_sheet_zoom_reset() {
+		}
+
+		// Assign in vector.
+		frames_ordered.set(index++, Pair<int, int>(order_by, idx));
```
Same thing with the increment.

YuriSizov reviewed (approved):
A couple of nitpicks, but overall the code looks good. The feature looks great!

AThousandShips commented:
Oh sorry will fix the merged commit one moment

AThousandShips commented:
Thank you! And I'm always happy to get good feedback

YuriSizov commented:
Thanks!

AThousandShips commented:
Thank you!
